apis/r/R/ephemeral.R

```diff
@@ -0,0 +1,450 @@
+      }
+      private$.ephemeral_error('custom', 'and have no context')
+    },
+    #' @field uri \dQuote{\code{ephemeral-collection:<MEMORY_ADDRESS>}}
+    uri = function(value) {
+      if (!missing(value)) {
+        private$.read_only_error('uri')
+      }
+      return(paste0('ephemeral-collection:', data.table::address(self)))
+    },
+    #' @field object \Sexpr[results=rd]{tiledbsoma:::rd_ephemeral_field()}
+    object = function(value) {
+      if (!missing(value)) {
+        private$.read_only_error('object')
+      }
+      private$.ephemeral_error('custom', 'and have no underlying object')
+    },
+    # Overwrite SOMACollectionBase fields
+    #' @field soma_type \Sexpr[results=rd]{tiledbsoma:::rd_ephemeral_field()}
+    soma_type = function(value) {
+      if (!missing(value)) {
+        private$.read_only_error('soma_type')
+      }
+      private$.ephemeral_error('custom', 'and have no SOMA type')
+    }
+  ),
+  private = list(
+    # Overwrite SOMACollectionBase private fields
+    tiledb_object = NULL,
+    tiledb_uri = NULL,
+    tiledb_platform_config = NULL,
+    .tiledbsoma_ctx = NULL,
+    mode = NULL,
+    # Overwrite TileDBGroup private fields
+    member_cache = NULL,
+    # Overwrite SOMACollectionBase private fields
+    soma_type_cache = NULL,
+    # Overwrite TileDBGroup private methods
+    open = function(mode) {
+      private$.ephemeral_error('opened')
+    },
+    initialize_object = function() {
+      private$.ephemeral_error('custom', 'and cannot be initialized')
+    },
+    get_all_members = function() {
+      if (!length(private$.data)) {
+        return(list())
+      }
+      members <- vector(mode = 'list', length = length(private$.data))
+      names(members) <- names(private$.data)
+      for (i in seq_along(members)) {
+        members[[i]] <- list(
+          type = tiledb::tiledb_object_type(private$.data[[i]]$uri),
+          uri = private$.data[[i]]$uri,
+          name = names(private$.data)[i]
+        )
+      }
+      return(members)
+    },
+    # Ephemeral fields
+    .data = NULL,
+    # Ephemeral methods
+    .ephemeral_error = function(type = 'added', msg = NULL) {
+      stopifnot("'type' must be a single character value" = is_scalar_character(type))
+      type <- match.arg(
+        arg = type,
+        choices = c(
+          'base',
+          'added',
+          'opened',
+          'edited',
+          'custom'
+        )
+      )
+      if (type == 'custom' && !is_scalar_character(msg)) {
+        stop("'msg' must be a single character value")
+      }
+      stop(
+        sQuote(self$class()),
+        " objects are ephemeral",
+        switch(
+          EXPR = type,
+          added = ' and cannot be added to',
+          opened = ' and cannot be opened',
+          edited = ' and cannot be edited',
+          custom = paste0(' ', trimws(msg))
+        ),
+        call. = FALSE
+      )
+    }
+  )
+)
+
+#' Virtual Collections
```

Comment:
as above -- i don't like virtual in r vs ephemeral in python -- without any (apparent) value-add in having different names